### PR TITLE
Fix integration tests for cmd state on Windows

### PR DESCRIPTION
**Pull request by twangboy**
### What does this PR do?

Fix integration tests for the cmd state on Windows.

https://github.com/saltstack/zh/issues/853

### CODE REVIEW

**Comment by cachedout** [comment]
This doesn't fix the problem. It just ignores it. Did you try explicitly closing the filehandle to see if that resolves the problem?

**Comment by twangboy** [comment]
@cachedout I did. It didn't make a difference.

**Comment by twangboy** [comment]
Go Go Jenkins!

**Comment by cachedout** [comment]
My concern here is two-fold:

a) This is fairly important functionality adjacent to the state system. Simply deciding not to test it is just piling on additional technical debt.

b) If a file that's being opened with `salt.utils.fopen` isn't being closed, ignoring that is potentially ignoring a _huge_ problem that will bite us down the road.

**Comment by terminalmage** [comment]
I couldn't agree more with @cachedout.

I think I see the problem here, and it is incorrect usage of `salt.utils.fopen()`. We should be using a ``with`` clause to ensure the filehandle is closed. @twangboy How did you test closing the filehandle, since none of these tests actually assign the filehandle to a variable?

Please try removing the ``skipIf`` from ``test_run_unless`` and edit it like so:

```python
    def test_run_unless(self):
        '''
        test cmd.run unless
        '''
        state_key = 'cmd_|-/var/log/messages_|-/var/log/messages_|-run'
        with salt.utils.fopen(self.state_file, 'w') as fp_:
            fp_.write(textwrap.dedent('''\
            /var/log/messages:
              cmd.run:
                - unless: echo cheese > {0}
            '''.format(self.test_file)))

        ret = self.run_function('state.sls', [self.state_name])
        self.assertTrue(ret[state_key]['result'])
```

**Comment by twangboy** [comment]
@terminalmage I assigned the handle to a variable and used the close function at the end....

The issue here isn't with this test, it is (IMO) a problem with the test suite. I get this error every time I run the test suite when it tries to remove the temporary test suite files. I believe them to be the same error.

**Comment by twangboy** [comment]
@terminalmage I tried using the `with` method... same result...

**Comment by terminalmage** [comment]
Oh, I misread... it's the ``temp_file`` that it can't remove. Of course it can't, it was opened with [tempfile.mkstemp](https://docs.python.org/2/library/tempfile.html#tempfile.mkstemp) and that leaves an open file _descriptor_, not a file _handle_.

I just opened https://github.com/twangboy/salt/pull/1 against your fork, merge it and try again, it should get rid of the failures. You should then be able to remove all of the skipIf logic.

With regard to the rest of the tests and the open file handling, I just opened a massive PR to fix them.

**Comment by twangboy** [comment]
@terminalmage Thanks for helping me get this one fixed!

**Comment by cachedout** [comment]
Almost there. :] Just a lint fix left: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/4101/violations/file/tests/integration/states/cmd.py/

**Comment by cachedout** [comment]
@twangboy This will no longer merge cleanly. Could you please rebase it? Thanks.

**Comment by twangboy** [comment]
@cachedout I think this one is ready, unless you want me to rebase it again...

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by cachedout** [comment]
Go Go Jenkins!

File: tests/integration/states/cmd.py [hunk 1]
```diff
@@ -63,12 +73,12 @@ def test_run_unless(self):
...
         state_key = 'cmd_|-/var/log/messages_|-/var/log/messages_|-run'
-        with salt.utils.fopen(self.state_file, 'w') as fp_:
-            fp_.write(textwrap.dedent('''\
-            /var/log/messages:
-              cmd.run:
-                - unless: echo cheese > {0}
-            '''.format(self.test_file)))
+        with salt.utils.fopen(self.state_file, 'w') as fb_:
+            fb_.write(textwrap.dedent('''
+                /var/log/messages:
```
Comment (cachedout) [inline tests/integration/states/cmd.py]: Did you want to modify this for Windows or no?

File: tests/integration/states/cmd.py [hunk 2]
```diff
@@ -63,12 +73,12 @@ def test_run_unless(self):
...
+                  cmd.run:
+                    - unless: echo cheese > {0}
```
Comment (cachedout) [inline tests/integration/states/cmd.py]: Does shell redirection work on Windows?

File: tests/integration/states/cmd.py [hunk 3]
```diff
@@ -77,13 +87,13 @@ def test_run_creates_exists(self):
...
-        with salt.utils.fopen(self.state_file, 'w') as fp_:
-            fp_.write(textwrap.dedent('''\
-            touch {0}:
-              cmd.run:
-                - creates: {0}
-            '''.format(self.test_file)))
+        state_key = 'cmd_|-echo >> {0}_|-echo >> {0}_|-run'.format(self.test_file)
+        with salt.utils.fopen(self.state_file, 'w') as fb_:
+            fb_.write(textwrap.dedent('''
+                echo >> {0}:
```
Comment (cachedout) [inline tests/integration/states/cmd.py]: Same question about shell redirection here.

File: tests/integration/states/cmd.py (hunk 1 above)
Comment (twangboy) [inline tests/integration/states/cmd.py]: This doesn't need to be modified for Windows.

File: tests/integration/states/cmd.py (hunk 2 above)
Comment (twangboy) [inline tests/integration/states/cmd.py]: Yes.

File: tests/integration/states/cmd.py (hunk 3 above)
Comment (twangboy) [inline tests/integration/states/cmd.py]: Yes. But there is no touch in Windows.

**Pull request merged**